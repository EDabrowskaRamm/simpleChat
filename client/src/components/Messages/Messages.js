import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'

import { MessagesRoot, MessagesWrapper } from './Messages.shards'

import ChatMessage from '../ChatMessage/ChatMessage'
import Input from '../Input/Input'

import { addMessage, updateMessage, showError } from '../../store/actions'

import { messageID, getMsgHour } from '../../utils/helpers'
import { URL, EVENT_TYPE } from '../../utils/consts.js'

const ConnectedMessages = ({
  getUser,
  getMessages,
  addMessage,
  updateMessage,
  showError
}) => {
  const [text, setText] = useState({
    name: '',
    message: '',
    id: messageID(),
    date: '',
    type: ''
  })
  const wsMsg = new WebSocket(URL)

  useEffect(() => {
    wsMsg.onopen = () => {
      console.log('connected')
    }

    wsMsg.onmessage = evt => {
      const socketData = JSON.parse(evt.data)

      switch (socketData.type) {
        case EVENT_TYPE.MESSAGE_EVENT:
          return addMessage(socketData)

        case EVENT_TYPE.MESSAGE_REMOVE:
          return updateMessage({
            ...socketData,
            message: 'Message removed',
            type: EVENT_TYPE.MESSAGE_REMOVE
          })

        case EVENT_TYPE.MESSAGE_EDIT:
          return updateMessage({
            ...socketData,
            type: EVENT_TYPE.MESSAGE_EDIT
          })

        default:
          return null
      }
    }

    wsMsg.onclose = () => {
      if (wsMsg.readyState === 3) {
        showError('WEBSOCKET DISCONNECTED')
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const handleChange = event => {
    setText({
      name: getUser.name,
      message: event.target.value,
      id: text.id,
      date: getMsgHour(),
      type: text.type
    })
  }

  const handleSubmit = event => {
    event.preventDefault()

    if (text.message.length) {

      if (text.type === EVENT_TYPE.MESSAGE_EDIT) {
        wsMsg.send(JSON.stringify({
          ...text,
          type: EVENT_TYPE.MESSAGE_EDIT
        }))
      } else {
        wsMsg.send(JSON.stringify({
          ...text,
          type: EVENT_TYPE.MESSAGE_EVENT
        }))
      }

      setText({
        name: getUser.name,
        message: '',
        id: messageID(),
        date: '',
        type: ''
      })
    }
  }

  const handleRemoveMessage = data => wsMsg.send(JSON.stringify({
    ...data,
    type: EVENT_TYPE.MESSAGE_REMOVE
  }))

  const handleEditMessage = data => {
    setText({
      name: getUser.name,
      message: data.message,
      id: data.id,
      date: data.date,
      type: EVENT_TYPE.MESSAGE_EDIT
    })
  }

  return <MessagesRoot>
    <MessagesWrapper>
      {getMessages.map(data => <ChatMessage
        key={data.id}
        messageInfo={data}
        isUser={getUser.name === data.name}
        removeMessage={() => handleRemoveMessage(data)}
        editMessage={() => handleEditMessage(data)}
      />)}
    </MessagesWrapper>
    <Input
      handleSubmit={handleSubmit}
      placeholder='Message'
      id='message'
      value={text.message}
      handleChange={handleChange}
    />
  </MessagesRoot>
}

const Messages = connect(
  state => ({
    getUser: state.user.details,
  }),
  dispatch => ({
    addMessage: (text) => dispatch(addMessage(text)),
    updateMessage: (id) => dispatch(updateMessage(id)),
    showError: (text) => dispatch(showError(text))
  })
)(ConnectedMessages)

export default Messages
