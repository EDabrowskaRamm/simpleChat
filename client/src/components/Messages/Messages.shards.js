import styled from '@emotion/styled'

export const MessagesRoot = styled.section`
  width: 100%;
  height: calc(100vh - 240px);
  display: flex;
  flex-direction: column;
`

export const MessagesWrapper = styled.div`
  height: 100%;
  padding: 0 30px;
  overflow: auto;
  flex-grow: 1;
`