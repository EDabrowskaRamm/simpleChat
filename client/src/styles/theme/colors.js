const _white = '#FFF'
const _colorGrey56 = '#868c97'
const _colorGrey61 = '#979c9f'
const _colorGrey63 = '#9a9fa8'
const _colorGrey89 = '#e1e3e5'
const _colorGrey92 = '#ebebeb'

const _pictonBlue = '#3d90f1'
const _oxfordBlue = '#354052'

export const background = {
  primary: _white,
  secondary: _colorGrey92,
}

export const text = {
  primary: _oxfordBlue,
  secondary: _colorGrey63,
  link: _pictonBlue,
  placeholder: _colorGrey61,
  left: _colorGrey56,
}

export const border = {
  input: _colorGrey56,
  chat: _colorGrey89,
  transparent: 'transparent',
}
