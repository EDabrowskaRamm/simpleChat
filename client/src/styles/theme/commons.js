import { border } from './colors'

export const borders = {
  main: `1px solid ${border.chat}`,
  input: `1px solid ${border.input}`,
}
