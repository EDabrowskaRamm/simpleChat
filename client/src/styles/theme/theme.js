import * as colors from './colors'
import * as commons from './commons'

export default {
  colors,
  commons
}
