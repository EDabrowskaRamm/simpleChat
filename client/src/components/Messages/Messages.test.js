import React from 'react'
import renderer from 'react-test-renderer'
import serializer from 'jest-emotion'
import { ThemeProvider } from 'emotion-theming'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

import Messages from './Messages'

import theme from '../../styles/theme/theme'
import initialState from '../../../__mocks__/fileMock'

expect.addSnapshotSerializer(serializer)

describe('renders Messages', () => {
  it('Snapshot', () => {
    const mockStore = configureStore()
    const mockData = {
      getMessages: initialState.messages.messagesList,
      getUser: initialState.user.details
    }

    const component = renderer.create(
      <Provider store={mockStore(initialState)}>
        <ThemeProvider theme={theme}>
          <Messages getMessages={mockData.getMessages} getUser={mockData.getUser} />
        </ThemeProvider>
      </Provider>)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
