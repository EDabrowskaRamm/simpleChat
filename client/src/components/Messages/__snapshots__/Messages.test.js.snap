// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Messages Snapshot 1`] = `
.emotion-12 {
  width: 100%;
  padding: 0 30px;
}

.emotion-4 {
  padding-left: 17px;
  padding-bottom: 17px;
}

.emotion-4:first-of-type {
  padding-top: 25px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 11px 0 5px;
}

.emotion-0 {
  font-weight: bold;
  text-transform: capitalize;
  -webkit-letter-spacing: 1.5px;
  -moz-letter-spacing: 1.5px;
  -ms-letter-spacing: 1.5px;
  letter-spacing: 1.5px;
  margin-right: 15px;
}

.emotion-1 {
  color: #979c9f;
}

.emotion-3 {
  -webkit-letter-spacing: 1.4px;
  -moz-letter-spacing: 1.4px;
  -ms-letter-spacing: 1.4px;
  letter-spacing: 1.4px;
  line-height: 1.6;
}

.emotion-11 {
  width: 100%;
  padding: 15px 0;
}

.emotion-10 {
  width: 100%;
  padding: 27px 33px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  border: 2px solid #868c97;
  border-radius: 9px;
}

.emotion-10:placeholder {
  color: #979c9f;
}

<section
  className="emotion-12"
>
  <div>
    <div
      className="emotion-4"
    >
      <div
        className="emotion-2"
      >
        <p
          className="emotion-0"
        >
          Jane Downey
        </p>
         
        <span
          className="emotion-1"
        >
          12:58:24 
        </span>
      </div>
      <p
        className="emotion-3"
      >
        After losing his wife seven years earlier, the eccentric Dr. John Dolittle, famed doctor and veterinarian of Queen Victoriaâ€™s England, hermits himself away behind the high walls of Dolittle Manor with only his menagerie of exotic animals for company.
      </p>
    </div>
    <div
      className="emotion-4"
    >
      <div
        className="emotion-2"
      >
        <p
          className="emotion-0"
        >
          Jane Downey
        </p>
         
        <span
          className="emotion-1"
        >
          12:59:54 
        </span>
      </div>
      <p
        className="emotion-3"
      >
        But when the young queen falls gravely ill, a reluctant Dolittle is forced to set sail on an epic adventure to a mythical island in search of a cure, regaining his wit and courage as he crosses old adversaries and discovers wondrous creatures.
      </p>
    </div>
  </div>
  <form
    className="emotion-11"
    onSubmit={[Function]}
  >
    <input
      className="emotion-10"
      id="message"
      onChange={[Function]}
      placeholder="Message"
      type="text"
      value=""
    />
  </form>
</section>
`;
